presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java

```diff
@@ -0,0 +1,311 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation;
+
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.block.BlockBuilder;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.type.BigintOperators;
+import com.facebook.presto.type.DoubleOperators;
+import com.facebook.presto.type.RealOperators;
+import org.openjdk.jol.info.ClassLayout;
+
+import static com.facebook.presto.common.type.DoubleType.DOUBLE;
+import static com.facebook.presto.common.type.RealType.REAL;
+import static com.facebook.presto.common.type.TypeUtils.isExactNumericType;
+import static com.facebook.presto.type.TypeUtils.expectedValueSize;
+import static java.lang.Math.max;
+import static java.util.Objects.requireNonNull;
+
+public class KeySetAndValues
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(KeySetAndValues.class).instanceSize();
+    private static final int EXPECTED_ENTRIES = 10;
+    private static final int EXPECTED_ENTRY_SIZE = 16;
+
+    private final Type keyType;
+    private final Type valueType;
+
+    private final TypedSet keySet;
+    private final Block valueBlock;
+    private final Block mapBlock;
+    private final Adder adder;
+
+    public KeySetAndValues(Type keyType, Type valueType, Block mapBlock)
+    {
+        this(keyType, valueType, null, null, mapBlock);
+    }
+
+    private KeySetAndValues(Type keyType, Type valueType, TypedSet keySet, Block valueBlock, Block mapBlock)
+    {
+        this.keyType = requireNonNull(keyType, "keyType is null");
+        this.valueType = requireNonNull(valueType, "valueType is null");
+        this.keySet = keySet;
+        this.valueBlock = valueBlock;
+        this.mapBlock = mapBlock;
+        this.adder = getAdder(valueType);
+    }
+
+    public Type getKeyType()
+    {
+        return keyType;
+    }
+
+    public void serialize(BlockBuilder out)
+    {
+        assert mapBlock != null != (keySet != null && valueBlock != null);
```

Comment:
This assert is too fancy. Can we just do `mapBlock != null && ...`

Reply:
Done